### PR TITLE
Fixed panic on concurrent index-header lazy load / unload

### PR DESCRIPTION
**Pull request by pracucci**
* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

We've found a panic that may occur when the index-header lazy loading is enabled. The panic happens because of an edge case happening between when the `load()` releases the write lock and re-acquires the read lock. We've been able to sistematically reproduce it with a unit test and this PR should fix it.

## Verification

Unit test.

### CODE REVIEW

**Review by pstibrany — APPROVED** [review]
LGTM, good check. Better to fail query with explicit error message than a panic.

File: pkg/block/indexheader/lazy_binary_reader.go [hunk 1]
```diff
@@ -23,6 +23,10 @@ import (
 	"github.com/thanos-io/thanos/pkg/objstore"
 )
 
+var (
+	errUnloadedWhileLoading = errors.New("the index-header has been concurrently unloaded")
```
Comment (MalloZup) [inline pkg/block/indexheader/lazy_binary_reader.go:27]: `nit` just my pov, do we need here a variable?

Can we just inline the error where it is used? I just checked the codebase it seems we use more the error directly.

But I might be wrong here. :grin:

**Comment by bwplotka** [comment]
Hmmm do we need this PR with your proper fix in separate PR @pracucci ? 🤔 

Or do we still want this to make sure we don't panic on races/bugs we did not yet discover? If that's so let's comment that this is unlikely, but just in case to not create Segfaults, WDYT?

**Comment by pracucci** [comment]
> Hmmm do we need this PR with your proper fix in separate PR @pracucci ? 🤔

I would say yes. This fix a race condition which still exists even if the other PR gets merged.

**Comment by pracucci** [comment]
@bwplotka I've rebased master and CI now passes. Do you still have concerns about this fix?

**Review by bwplotka — APPROVED** [review]
I don't have time to review it properly TBH, but don't want to block you. 

Not sure if we really have to  fail whole query (it's enough to just retry kind of 🤔 )

**Comment by bwplotka** [comment]
It's LGTM for a start (:

**Comment by bwplotka** [comment]
Or rather: I reviewed it properly and it makes sense for now, but just don't have time to suggest something better 🙃
